spec/blocksync/verification.md

```diff
@@ -0,0 +1,120 @@
+
+- The block is within the trusting period, we can punish validators, and provide guarantees that if a block is signed by trusted validators, they are indeed correct. 
+- The block is outside the trusting period and we cannot guarantee that the validators are still honest. We can however make a potential attack by those validators more complicated by additionally verifying the block using other peers as *witnesses*.
+
+A state becomes trusted once it passes the validation. The section below discusses this state in more details.
+
+### Trusted state
+
+Currently there are two possible ways to obtain a trusted state :
+ 1. If the node is blocksync-ing from genesis, we have to verify the very first block and mark it as the initial trusted state.
+ 
+ [//]: # (trust the validators provided in the initial state and use them to verify the initial block received from peers.)
+
+ 2. The node loads the last block stored in its local block store. 
+
+**Case 1 - blocksync from genesis**
+
+In this case we only verify that the validator hash of the block matches the validator set of the initial state. We will use the initially provided validator set for verification and further verify the block against witne+sses. In this scenario, this state is very likely to be outside the trusting period. We will accept this block as a trusted state and store it inside the node's block store. 
+
+It is worth noting that, running block sync from the first height is significantly slower than running statesync first. However, statesync does not keep the entire blockchain history and some operators might opt not to state sync. The reason is that, if sufficiently many nodes state sync and other nodes who have historical data fail or leave the network, we have no history.
+
+**Case 2 - blocksync starting from pre-existing state**
+
+If the node is not starting from gensis, there will be pre-existing state in its local block store. 
+
+This can happen in the following two cases:
+
+- The node is blocksyncing after completing state sync.
+
+- The node has crashed and is now recovering.
+
+In both of these cases, we can trust the state in the store. State sync runs the light client verification protocol. If the crash happened during consensus or sync-ing, the only blocks in the store are blocks that are verified or decided. 
+
+If we can trust this initial state at height H, we can use this fact to eliminate faulty blocks whose `lastCommit` at height H+1 does not verify the trusted block. Additionally, we can double check the validator sets that signed the block at H + 1 against the expected validator sets for this height. The expected validator sets are stored within the trusted state and returned to us upon applying the block and executing it against the application.
```

Comment:
What if the initial height H has fallen out of the trusting period?

Reply:
The initial (trusted) state (maybe the wording should be adjusted) is a block and the commit that validated this  block. We use it to increase the trust in the block at H + 1 but we do not have guarantees that the validators that signed block at H are indeed still honest.

Reply:
Good point, I think we should discuss this more. There were a few solutions proposed to deal with this but a lot of them were dismissed due to the tradeoff between added complexity and performance + likelyhood of this happening. 

Bottom line, all scenarios in which a syncing node is fooled will not impact o ther nodes performing consensus because once the node switches to consensus, the faulty blocks will be detected. However, as you stated, the syncing node can get punished even though it had no malicious intent.  This scenario is mentioned below in line 107.

Reply:
While in blocksync, the node is not supposed to interact with other nodes (other than just requesting blocks), so we're safe. The problem is if we switch to consensus (e.g. due to lack of progress) and then start voting, or proposing...
OTOH, checking that we have reached the trusting window doesn't seem complex or expensive, but I might be missing something.